### PR TITLE
Fix Orphaned ACL race condition

### PR DESCRIPTION
**Pull request by astoycos**
There is a race condition, described here
(https://bugzilla.redhat.com/show_bug.cgi?id=1859682)
that results in ACL's and their associated network
entities being orphaned upon the deletion of
a namespace

Specifically, if a namespace is deleted by OVN-K8's
prior to all the Network Policies being handled,
the ACLS, portgroups, address sets, and handlers
become orphaned in `nbdb` beacuse a namespace lock
can never be acquired in `policy.go`'s
`deleteNetworkPolicy()`

This fix ensures that all network entities involving
a namespace's NetworkPolicies are deleted prior to the
namespace's deletion

TODO: I think this is just a quick fix, ultimately
It would be nice to decouple the NetworkPolicy's
dependency on namespaces within OVN maybe by moving the
[`networkPolicies` map](https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/ovn.go#L79)
out of the `namespace` object to the
[`controller` object](https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/ovn.go#L79)? 

signed-off-by: Andrew Stoycos <astoycos@redhat.com>

**- Special notes for reviewers**

**- How to verify it**

To reproduce this race add a sleep for approximately 3 seconds [here](https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/policy.go#L749) 

Then bring up an `ovn-k8's` cluster locally using kind and your custom OVN build. 

Next make a new namespace titled `test1` and create 10 `NetworkPolicies` in that namespace 

Finally delete the namespace and look in the `nbdb` for any orphaned ACLs, port-groups, and address sets 

There will be orphaned ACLs attatched to a NS that doesn't exist as shown below 

```
[astoycos@localhost ovn-org]$ ./list-acls.sh 
Current ACLs
ovnkube-master-7888658d8-67nqz:
Defaulting container name to ovn-northd.
Use 'kubectl describe pod/ovnkube-master-7888658d8-67nqz -n ovn-kubernetes' to see all of the containers in this pod.
external_ids        : {Ingress_num="0", ipblock_cidr="false", l4Match=None, namespace=test2, policy=allow-from-blue-4, policy_type=Ingress}
external_ids        : {Ingress_num="0", ipblock_cidr="false", l4Match=None, namespace=test2, policy=allow-from-blue-2, policy_type=Ingress}
external_ids        : {Ingress_num="0", ipblock_cidr="false", l4Match=None, namespace=test2, policy=allow-from-blue-10, policy_type=Ingress}
external_ids        : {Ingress_num="0", ipblock_cidr="false", l4Match=None, namespace=test2, policy=allow-from-blue-1, policy_type=Ingress}
external_ids        : {Ingress_num="0", ipblock_cidr="false", l4Match=None, namespace=test2, policy=allow-from-blue-9, policy_type=Ingress}
external_ids        : {Ingress_num="0", ipblock_cidr="false", l4Match=None, namespace=test2, policy=allow-from-blue-3, policy_type=Ingress}
external_ids        : {Ingress_num="0", ipblock_cidr="false", l4Match=None, namespace=test2, policy=allow-from-blue-5, policy_type=Ingress}
external_ids        : {Ingress_num="0", ipblock_cidr="false", l4Match=None, namespace=test2, policy=allow-from-blue-6, policy_type=Ingress}
[astoycos@localhost ovn-org]$ oc get namespaces
NAME                 STATUS   AGE
default              Active   31m
kube-node-lease      Active   31m
kube-public          Active   31m
kube-system          Active   31m
local-path-storage   Active   31m
ovn-kubernetes       Active   27m
[astoycos@localhost ovn-org]$ 
```

**- Description for the changelog**

This duplicates the deletion of network entities found in `go-contoller/pkg/ovn/policy.go` to be included upon 
namespace deletion in a new `go-contoller/pkg/ovn/namespace.go` function `cleanNPentities` that is shared and called in both `namespace.go` and `policy.go`. Also I modified `localPodDelDafaultDeny` to take a `namespacePolicy` object rather than a `knet.NetworkPolicy` since in the case of this issue, we need to confirm that everything involving NP's can be cleaned up by both `policy.go` and `namespace.go`.

### CODE REVIEW

**Comment by astoycos** [comment]
@JacobTanenbaum

**Comment by trozet** [comment]
@astoycos nice find! I think the fix you have for now is just fine. However, I think there is another place this can happen, in deleteLogicalPort on the pod side, we call:

```
	if err := oc.deletePodFromNamespace(pod.Namespace, portInfo); err != nil {
		klog.Errorf(err.Error())
	}
```

which is:
```
func (oc *Controller) deletePodFromNamespace(ns string, portInfo *lpInfo) error {
	nsInfo := oc.getNamespaceLocked(ns)
	if nsInfo == nil {
		return nil
	}
	defer nsInfo.Unlock()

	if err := nsInfo.addressSet.DeleteIPs(createIPAddressSlice(portInfo.ips)); err != nil {
		return err
	}

	// Remove the port from the multicast allow policy.
	if oc.multicastSupport && nsInfo.multicastEnabled {
		if err := podDeleteAllowMulticastPolicy(ns, portInfo); err != nil {
			return err
		}
	}

	return nil
}

If we fail to get namespace there because namespace event already came and deleted it, then we never get to podDeleteAllowMulticastPolicy. Which means we never remove the port from that port group. Maybe it doesn't matter because when the port gets deleted OVN will automatically remove it from the port group? Do you mind testing this path as well?
```

**Comment by astoycos** [comment]
@trozet sounds good I will test this as well, It definitely seems like it could still be an issue especially If OVN still sees the Pod's IP in the address set.

File: go-controller/pkg/ovn/namespace.go [hunk 1]
```diff
@@ -331,6 +331,30 @@ func (oc *Controller) deleteNamespace(ns *kapi.Namespace) {
 		return
 	}
 	defer nsInfo.Unlock()
+
+	klog.V(5).Infof("Deleting Namespaces's NetworkPolicy entities")
```
Comment (JacobTanenbaum) [inline go-controller/pkg/ovn/namespace.go:335]: I suspect there is more cleanup that will need to be done for namespace cleanup, could you wrap this in a function like `(oc *Controller) cleanUpNsinfo()`

Comment (astoycos) [reply go-controller/pkg/ovn/namespace.go:335]: ack... I agree, this was more of a prototype to make sure it stopped the race.  What else do you think needs to be cleaned up for NPs? I basically mimicked the functionality of `deleteNetworkPolicy` minus `oc.localPodDelDefaultDeny(policy, portInfo)` which based on my understanding shouldn't matter here since all the Pods in the namespace will be deleted

**Comment by astoycos** [comment]
> @astoycos nice find! I think the fix you have for now is just fine. However, I think there is another place this can happen, in deleteLogicalPort on the pod side, we call:
> 
> ```
> 	if err := oc.deletePodFromNamespace(pod.Namespace, portInfo); err != nil {
> 		klog.Errorf(err.Error())
> 	}
> ```
> 
> which is:
> 
> ```
> func (oc *Controller) deletePodFromNamespace(ns string, portInfo *lpInfo) error {
> 	nsInfo := oc.getNamespaceLocked(ns)
> 	if nsInfo == nil {
> 		return nil
> 	}
> 	defer nsInfo.Unlock()
> 
> 	if err := nsInfo.addressSet.DeleteIPs(createIPAddressSlice(portInfo.ips)); err != nil {
> 		return err
> 	}
> 
> 	// Remove the port from the multicast allow policy.
> 	if oc.multicastSupport && nsInfo.multicastEnabled {
> 		if err := podDeleteAllowMulticastPolicy(ns, portInfo); err != nil {
> 			return err
> 		}
> 	}
> 
> 	return nil
> }
> ```
> 
> If we fail to get namespace there because namespace event already came and deleted it, then we never get to podDeleteAllowMulticastPolicy. Which means we never remove the port from that port group. Maybe it doesn't matter because when the port gets deleted OVN will automatically remove it from the port group? Do you mind testing this path as well?

so @trozet  It seems in this case there is not an issue because 

1. Even If I force a race and the Pod's address is not removed from the NS's address set, the address set is ultimately deleted during namespace deletion anyways. 

and 

2. Even if the pod's port is not removed from the multicast allow policy all the policies will be ultimately cleaned up correctly when the namespace is deleted thanks to this PR  

And those are the only two things that are in risk of not happening if a namespace lock is never acquired [here](https://github.com/ovn-org/ovn-kubernetes/blob/59af27808d619939f5ee58a60a70d98862006574/go-controller/pkg/ovn/namespace.go#L71)

File: go-controller/pkg/ovn/namespace.go (hunk 1 above)
Comment (astoycos) [reply go-controller/pkg/ovn/namespace.go]: fixed

File: go-controller/pkg/ovn/namespace.go [hunk 2]
```diff
@@ -425,3 +427,31 @@ func (oc *Controller) deleteNamespaceLocked(ns string) *namespaceInfo {
 
 	return nsInfo
 }
+
+func (oc *Controller) cleanUpNsinfo(nsInfo *namespaceInfo) {
+
+	klog.V(5).Infof("Deleting Namespaces's NetworkPolicy entities")
+
+	for _, np := range nsInfo.networkPolicies {
+		delete(nsInfo.networkPolicies, np.name)
```
Comment (dcbw) [inline go-controller/pkg/ovn/namespace.go:436]: Can we not share code with deleteNetworkPolicy()?

Comment (astoycos) [reply go-controller/pkg/ovn/namespace.go:436]: So `deleteNetworkPolicy()` is called with a `*knet.NetworkPolicy` which we don't have access to within the `namespace`  object (all we have is the Namespace stored *namespacePolicy), and `deleteNetworkPolicy()` works great except for this edge case where the namespace object manages to delete before the policy elements.  So therefore It would be harder to share and refactor the the code in `deleteNetworkPolicy` rather than add a temporary addendum to `namespace.go` I believe.  Ultimately I don't think we should have network policies be dependent on namespaces in OVN since we receive the events for deletion of both entities asynchronously.

Comment (astoycos) [reply go-controller/pkg/ovn/namespace.go:436]: @dcbw Unless you have a suggestion that I'm not understanding :)

Comment (JacobTanenbaum) [reply go-controller/pkg/ovn/namespace.go:436]: they both use the namespacePolicy struct that is part of nsInfo...  you could define `cleanUpNamespacePolicy(np *namespacePolicy)` in policy.go and call it from both `cleanUpNsinfo()` and `deleteNamespace()`

Comment (astoycos) [reply go-controller/pkg/ovn/namespace.go:436]: ok I see, thanks updating the patch now

**Comment by astoycos** [comment]
> /retest

@JacobTanenbaum Trying to figure out if that was a flake or I broke something

**Comment by JacobTanenbaum** [comment]
i think when the mocking fails it is a flake

File: go-controller/pkg/ovn/namespace.go [hunk 3]
```diff
@@ -331,8 +331,17 @@ func (oc *Controller) deleteNamespace(ns *kapi.Namespace) {
 		return
 	}
 	defer nsInfo.Unlock()
+
+	klog.V(5).Infof("Deleting Namespaces's NetworkPolicy entities")
+	for _, np := range nsInfo.networkPolicies {
+		delete(nsInfo.networkPolicies, np.name)
+		np.deleted = true
+
```
Comment (JacobTanenbaum) [inline go-controller/pkg/ovn/namespace.go:339]: nit whitespace

**Comment by astoycos** [comment]
@trozet @dcbw This should be about ready to go

**Comment by astoycos** [comment]
@danwinship

**Comment by astoycos** [comment]
@danwinship @dcbw Just need you look over/ approval, thanks in advance!

File: go-controller/pkg/ovn/policy.go [hunk 4]
```diff
@@ -332,7 +332,6 @@ func createDefaultDenyMulticastPolicy() error {
 		return fmt.Errorf("failed to create port_group for %s (%v)",
 			portGroupName, err)
 	}
-
```
Comment (dcbw) [inline go-controller/pkg/ovn/policy.go:335]: Seems unrelated?

File: go-controller/pkg/ovn/policy.go [hunk 5]
```diff
@@ -747,21 +746,26 @@ func (oc *Controller) deleteNetworkPolicy(policy *knet.NetworkPolicy) {
...
 	for _, portInfo := range np.localPods {
-		oc.localPodDelDefaultDeny(policy, portInfo)
+		oc.localPodDelDefaultDeny(np, portInfo)
 	}
 
 	// Delete the port group
 	deletePortGroup(np.portGroupName)
 
-	// Delete ingress/egress address sets
+	// Delete ingress/ egress address sets
```
Comment (dcbw) [inline go-controller/pkg/ovn/policy.go:768]: Unecessary whitespace addition after the / ?

File: go-controller/pkg/ovn/policy.go [hunk 6]
```diff
@@ -747,21 +746,26 @@ func (oc *Controller) deleteNetworkPolicy(policy *knet.NetworkPolicy) {
...
+			policy.Name, policy.Namespace)
 		return
 	}
 	defer np.Unlock()
 
-	// We should now stop all the handlers go routines.
+	oc.cleanNPentities(np)
+}
+
+func (oc *Controller) cleanNPentities(np *namespacePolicy) {
```
Comment (dcbw) [inline go-controller/pkg/ovn/policy.go:758]: @astoycos this is really a function to clean up a single namespacePolicy, so how about `destroyNamespacePolicy()`?

File: go-controller/pkg/ovn/namespace.go [hunk 7]
```diff
@@ -331,8 +331,17 @@ func (oc *Controller) deleteNamespace(ns *kapi.Namespace) {
 		return
 	}
 	defer nsInfo.Unlock()
+
+	klog.V(5).Infof("Deleting Namespaces's NetworkPolicy entities")
+	for _, np := range nsInfo.networkPolicies {
+		delete(nsInfo.networkPolicies, np.name)
+		np.deleted = true
```
Comment (dcbw) [inline go-controller/pkg/ovn/namespace.go:338]: I think we can move this `np.deleted = true` to cleanNPentities/destroyNamespacePolicy()? That way it only needs to be set in one place. Both places that set it call the clean/destroy function right after.

File: go-controller/pkg/ovn/namespace.go [hunk 8]
```diff
@@ -331,8 +331,17 @@ func (oc *Controller) deleteNamespace(ns *kapi.Namespace) {
...
+		oc.cleanNPentities(np)
+	}
+
 	oc.deleteGWRoutesForNamespace(nsInfo)
 	oc.multicastDeleteNamespace(ns, nsInfo)
+
```
Comment (dcbw) [inline go-controller/pkg/ovn/namespace.go:344]: Unecessary newline.

File: go-controller/pkg/ovn/namespace.go (hunk 7 above)
Comment (astoycos) [reply go-controller/pkg/ovn/namespace.go]: Yep I agree @dcbw should be all set now

File: go-controller/pkg/ovn/policy.go (hunk 6 above)
Comment (astoycos) [reply go-controller/pkg/ovn/policy.go]: Yeah I couldn't come up with a good name... I like this suggestion

**Comment by trozet** [comment]
@aojea should we be testing network latency tests in our CI?
[Fail] [sig-network] Service endpoints latency [It] should not be very high  [Conformance] 
/home/runner/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/service_latency.go:129

Wondering if we don't have enough resources to handle those.

**Comment by danwinship** [comment]
The test creates 200 pods and requires that half of them are created in less than 20 seconds and 99% of them are created in less than 50 seconds. We got:

```
2020-09-18T22:24:14.8346335Z Sep 18 22:23:47.273: INFO: 50 %ile: 13.748112881s
2020-09-18T22:24:14.8347819Z Sep 18 22:23:47.273: INFO: 90 %ile: 25.017899337s
2020-09-18T22:24:14.8348174Z Sep 18 22:23:47.273: INFO: 99 %ile: 1m2.258993237s
2020-09-18T22:24:14.8348550Z Sep 18 22:23:47.273: INFO: Total sample count: 200
2020-09-18T22:24:14.8390653Z Sep 18 22:23:47.273: FAIL: Tail (99 percentile) latency should be less than 50s
```

And it wasn't a single outlier; there was a pretty even distribution of times. Looks like yeah, maybe with our current level of parallelism we're hitting problems.

But it's good to have some performance sanity-checking... Maybe the test should be made configurable so you can tune it better to your test environment.

**Comment by aojea** [comment]
> @aojea should we be testing network latency tests in our CI?
> [Fail] [sig-network] Service endpoints latency [It] should not be very high [Conformance]
> /home/runner/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/service_latency.go:129
> 
> Wondering if we don't have enough resources to handle those.

I will keep it, unless we see that is a common failure, it may that we are taking too long in bind the pods or provide the IP.

Let's keep an eye on it, however, this is not a test that I see failing commonly

**Comment by astoycos** [comment]
@dcbw I addressed your changes and this should be ready to go

File: go-controller/pkg/ovn/namespace.go [hunk 9]
```diff
@@ -331,6 +331,12 @@ func (oc *Controller) deleteNamespace(ns *kapi.Namespace) {
 		return
 	}
 	defer nsInfo.Unlock()
+
+	klog.V(5).Infof("Deleting Namespaces's NetworkPolicy entities")
```
Comment (knobunc) [inline go-controller/pkg/ovn/namespace.go:335]: Typo: Is this a single namespace?  The message should either be Namespace's or if it does multiple Namespaces'

Comment (astoycos) [reply go-controller/pkg/ovn/namespace.go:335]: It's a single namespace, thanks for catching that!

**Comment by JacobTanenbaum** [comment]
lgtm

**Comment by dcbw** [comment]
/lgtm
failures are `should be able to preserve UDP traffic when server pod cycles for a ClusterIP service` and `service endpoint latency should not be high` which are not caused by this PR.

**Comment by trozet** [comment]
@vpickard note you are looking into the above failure that we keep seeing flake in downstream:
preserve UDP traffic when server pod cycles for a ClusterIP service and service endpoint latency should not be high which are not caused by this PR.

Looks like we hit it in upstream.

**Comment by vpickard** [comment]
I chatted some with @aojea this morning about running this test. If I run this test manually on my local KIND 0.9.0 setup, it passes. Running the test like this: 

`./_output/local/go/bin/e2e.test -kubeconfig  $HOME/admin.conf -ginkgo.focus="should be able to preserve UDP traffic when server pod cycles for a ClusterIP service"
`

If I run the test like this, I see the failure:
`PARALLEL=true make -C test shard-network`

Antonio pointed me to a k8s issue about flaky tests, which was very helpful! Thanks Antonio. Here is the link:
https://github.com/ovn-org/ovn-kubernetes/issues/1707

From here, I did a search on upstream k8s issues, and see that this test is also flaking in upstream k8s:
https://storage.googleapis.com/k8s-gubernator/triage/index.html?test=conntrack 

So, I will continue to follow the suggestions in the links shared, and debug to figure out why this is flaking.

**Comment by aojea** [comment]
> preserve UDP traffic when server pod cycles for a ClusterIP service

I have the upstream bug and was investigating it for a while, improving it's success rate, but UDP is lossy by definition and I don't think that this test in a noise environment can achieve 100% of success rate, but 95% should be doable. 
Some of the issues with this test was:
- conntrack stale entries, that's exactly what the test looks for. Kube-proxy has a different logic than ovn, one place to check is if we are deleting the stale entries when the pod is deleted correctly.
- enpoints rules programmed in the dataplane. Ff the pod is deleted but the new pod is not binded in time, the endpoint publised and the service -> new endpoint rules are not in place (in OVN is a load balancer) in a timely manner, it will fail.

I will try to time those events and try to understand if ovn is lagging in some of them,
